### PR TITLE
Improving dapps API test coverage - Closes #560

### LINKED ISSUE #560: Remove selected dapps functionalties, pending rewrite
For Lisk `1.0.0`, we will remove the following selected dapp related functionalities, pending a full rewrite.

**Functionalities to be removed:**

- Download
- Installation
- Uninstallation
- Process management (and sandboxing)
- Interprocess communication
- Peer to peer communication

**Functionalities to be kept:**

- Application registration
   (but only using `lisk-js`, therefore `PUT /dapps` will be removed)
- Application deposits
   (but only using `lisk-js`, therefore `PUT /dapps/transaction` will be removed)
- Application withdrawals
   (but only using `lisk-js`, therefore `PUT /dapps/withdrawal` will be removed)
- Getting an individual application registration
- Listing of registered applications
- Listing of application categories

Related issues:

- https://github.com/LiskHQ/lisk/issues/225
- https://github.com/LiskHQ/lisk/issues/457
- https://github.com/LiskHQ/lisk-js/issues/106
- https://github.com/LiskHQ/lisk-build/issues/160

### PR DESCRIPTION
**Pull request by karmacoma**
Improved dapps API test coverage of remaining functionalities, after selected removal in: https://github.com/LiskHQ/lisk/pull/655

Covering:

- `GET /api/dapps/get?id=`
- `GET /api/dapps/categories`
- `GET /api/dapps/?type=`
- `GET /api/dapps/?name=`
- `GET /api/dapps/?category=`
- `GET /api/dapps/?link=`
- `GET /api/dapps/?limit=`
- `GET /api/dapps/?limit=1&offset=`
- `GET /api/dapps/?orderBy=`

Fixing:

- Reverted premature dapps API change from `/:id` back to `/get`.
- Fixed incorrect error message yielded by `DApps.prototype.internal.list`.

Closes #560

### CODE REVIEW

File: modules/dapps.js [hunk 1]
```diff
@@ -129,7 +129,8 @@ __private.list = function (filter, cb) {
 
 	if (filter.category) {
 		var category = dappCategories[filter.category];
-		if (category != null && category !== undefined) {
+
+		if (category != null) {
```
Comment (4miners) [inline modules/dapps.js]: Why removed check for `undefined`?

File: test/api/dapps.js [hunk 2]
```diff
@@ -18,214 +22,427 @@ before(function (done) {
...
+before(function (done) {
+	var account = node.randomAccount();
+	var options = node.blockDataDapp;
+	var transaction = node.lisk.dapp.createDapp(node.gAccount.password, null, options);
+	node.blockDataDapp.transactionId = transaction.id;
+
+	postTransaction(transaction, done);
+});
+
+before(function (done) {
```
Comment (4miners) [inline test/api/dapps.js]: Can we join those 2 `before` into one?

File: test/api/dapps.js [hunk 3]
```diff
@@ -18,214 +22,427 @@ before(function (done) {
...
+
+	it('using id with length > 20 should fail', function (done) {
+		getDapp('012345678901234567890', function (err, res) {
+			node.expect(res.body).to.have.property('success').to.not.be.ok;
+			node.expect(res.body).to.have.property('error').that.is.equal('String is too long (21 chars), maximum 20: #/id');
+			done();
+		});
+	});
+
+	it('using unknown id should fail', function (done) {
```
Comment (4miners) [inline test/api/dapps.js]: Should be tested also using it that is not numeric.

File: test/api/dapps.js [hunk 4]
```diff
@@ -18,214 +22,427 @@ before(function (done) {
...
-				node.expect(res.body.dapps[0].category).to.equal(node.dappCategories[randomCategory]);
-			}
+			node.expect(res.body).to.have.property('dapps').that.is.an('array').and.has.lengthOf(2);
+			node.expect(res.body.dapps[0].type).to.equal(0);
+			node.expect(res.body.dapps[1].type).to.equal(0);
+			done();
+		});
+	});
+
+	it('using type == 1 should be ok', function (done) {
```
Comment (4miners) [inline test/api/dapps.js]: Same - not numeric type should be tested.

File: test/api/dapps.js [hunk 5]
```diff
@@ -18,214 +22,427 @@ before(function (done) {
...
-				node.expect(res.body.categories[i]).to.equal(node.dappCategories[i]);
-			}
+			node.expect(res.body).to.have.property('dapps').that.is.an('array');
+			node.expect(res.body.dapps[0].category).to.equal(0);
+			node.expect(res.body.dapps[1].category).to.equal(1);
+			done();
+		});
+	});
+
+	it('using orderBy == "unknown:unknown" should be ok', function (done) {
```
Comment (4miners) [inline test/api/dapps.js]: `shoud fail`

File: modules/dapps.js (hunk 1 above)
Comment (karmacoma) [inline modules/dapps.js]: Because:

```
> null != undefined
false
> null != null
false
> null != true
true
```

File: test/api/dapps.js (hunk 2 above)
Comment (karmacoma) [inline test/api/dapps.js]: I did like this to separate fixture setup A from B. There is similar pattern in other tests.

File: test/api/dapps.js (hunk 3 above)
Comment (karmacoma) [inline test/api/dapps.js]: True, will add.

File: test/api/dapps.js (hunk 4 above)
Comment (karmacoma) [inline test/api/dapps.js]: Ok thanks.

File: test/api/dapps.js (hunk 5 above)
Comment (karmacoma) [inline test/api/dapps.js]: Yes should fail. I covered the code as it is currently.

File: test/api/dapps.js (hunk 2 above)
Comment (karmacoma) [inline test/api/dapps.js]: Refactoring it.

File: modules/dapps.js (hunk 1 above)
Comment (4miners) [inline modules/dapps.js]: Yeah, right.

**Comment by Isabello** [comment]
Looks great. Thank you for your hard work @karmacoma

**Pull request merged**